src/middlewared/middlewared/plugins/jbof/crud.py

```diff
@@ -0,0 +1,546 @@
+        """
+        verrors, data = await self.validate(data, 'jbof_create')
+        verrors.check()
+
+        mgmt_ip = data['mgmt_ip1']
+        shelf_index = data['index']
+
+        # Everything looks good so far.  Attempt to hardwire the dataplane.
+        try:
+            await self.middleware.call('jbof.hardwire_dataplane', mgmt_ip, shelf_index, 'jbof_create.mgmt_ip1', verrors)
+            if verrors:
+                await self.middleware.call('jbof.unwire_dataplane', mgmt_ip, shelf_index)
+        except Exception as e:
+            self.logger.error('Failed to add JBOF', exc_info=True)
+            # Try a cleanup
+            try:
+                await self.middleware.call('jbof.unwire_dataplane', mgmt_ip, shelf_index)
+            except Exception:
+                pass
+            verrors.add('jbof_create.mgmt_ip1', f'Failed to add JBOF: {e}')
+        verrors.check()
+
+        data['id'] = await self.middleware.call(
+            'datastore.insert', self._config.datastore, data,
+            {'prefix': self._config.datastore_prefix})
+
+        return await self.get_instance(data['id'])
+
+    @accepts(
+        Int('id', required=True),
+        Patch(
+            'jbof_create', 'jbof_update',
+            ('attr', {'update': True})
+        )
+    )
+    async def do_update(self, id_, data):
+        """
+        Update JBOF of `id`
+        """
+        old = await self.get_instance(id_)
+        new = old.copy()
+        new.update(data)
+        verrors, data = await self.validate(new, 'jbof_update', old)
+        verrors.check()
+
+        if old['uuid'] != new['uuid']:
+            self.middleware.logger.debug('Changed UUID of JBOF')
+            await self.middleware.call('jbof.unwire_dataplane', old['mgmt_ip1'], old['index'])
+            await self.middleware.call('jbof.hardwire_dataplane', new['mgmt_ip1'], new['index'], 'jbof_update.mgmt_ip1', verrors)
+
+        await self.middleware.call(
+            'datastore.update', self._config.datastore, id_, new,
+            {'prefix': self._config.datastore_prefix})
+
+        return await self.get_instance(id_)
+
+    @accepts(Int('id'))
+    async def do_delete(self, id_):
+        """
+        Delete a JBOF by ID.
+        """
+        # Will make a best-effort un tear down existing connections / wiring
+        # To do that we first need to fetch the config.
+        data = await self.get_instance(id_)
+        await self.middleware.run_in_thread(self.ensure_redfish_client_cached,
+                                            data['mgmt_ip1'],
+                                            data.get('mgmt_username'),
+                                            data.get('mgmt_password'))
+        try:
+            await self.middleware.call('jbof.unwire_dataplane', data['mgmt_ip1'], data['index'])
+        except Exception:
+            self.middleware.logger.debug('Unable to unwire JBOF @%r', data['mgmt_ip1'])
```

Comment:
Depending on where the failure occurred it could mean:
- nvme disks don't get cleanly removed (very unlikely that everything fails, so they'll possibly get _uncleanly_ removed)
- host RDMA IPs don't get torn down
- ES24N networking doesn't get torn down.

I suppose the best clean up would be a reboot.

Reply:
Maybe generate an alert then that reboot may be required due to potentially incomplete teardown?

Reply:
Added `JBOFTearDownFailureAlertClass`